gnetlist/src/gnetlist.1
.TH gEDA 1 "July 5th, 1999" Version 19990705
.SH NAME
gnetlist - GPL Netlist extraction/generation
.SH SYNOPSIS
.B gnetlist
[-i] [-v] [-r rcfilename] [-g guile_procedure] [-o output_filename] schematic1 [... schematicN]
.SH DESCRIPTION
.PP
\fIgnetlist\fP is the netlist extraction/generation program which is part gEDA 
(GPL Electronic Design Automation) toolset.  This program takes a schematic
for its input and outputs a netlist.  As of the writing of this man page
a test netlist format is the only existing output format ("ales-netlist").  

gnetlist depends heavily on guile (a scheme based scripting language).  It 
uses guile to define the output format.  Basically gnetlist reads a schematic,
creates an internal representation of the various connections, and then a
guile script extracts the connections into some netlist format.  

gnetlist is very much so a work in progress.  Currently it supports the
following backends: 

        1) Allegro netlist format (-g allegro)
        2) BAE netlist format (-g bae)
        3) BOM / BOM2 - Bill of Materials (-g bom and -g bom2)
        4) DRC - Start of a design rule checker (-g drc)
        5) gEDA - native format, mainly used for testing (-g geda)
        6) Gossip netlist format (-g gossip)
        7) PADS netlist format (-g pads)
        8) PCB / PCBboard (-g PCB and -g PCBboard)
        9) ProtelII netlist format (-g protelII)
        10) Spice compatible netlist format (-g spice)
        11) Enhanced spice compatible netlist format (-g spice-sdb)
        12) Switcap netlist format (-g switcap)
        13) Tango netlist format (-g tango)
        14) Verilog code (-g verilog)
        15) VHDL code (-g vhdl)
        16) VIPEC netlist format (-g vipec)

For more info on these formats please look at the README.*

Please read the official (non-existant) documentation on how to use
gnetlist, since this man page just describes the command line arguments
and a few examples on how to run gnetlist.

.SH OPTIONS
.l
\fIgnelist\fP accepts the following options:
.TP 8
.B -q
Quiet mode on.  This mode turns off all warnings/notes/messages. (optional)
.TP 8
.B -v 
Verbose mode on.  This mode gives as much feedback to the user as possible. (optional)
.TP 8
.B -g guile_procedure
Specify the guile procedure which is executed to create the netlist.  
.TP 8
.B -o output_filename
Specify the filename which will contain the netlist generated by gnetlist.  If this option is not specified the default filename is "output.net".
.TP 8
.B -c string
Pass the specified string to the guile interpreter.  This allows you to 
execute arbitrary guile scripts from the command line.  Be sure to surround
the string with either single or double quotes to satisfy your shell.  The
string is execute before any init or netlist backend scheme code is
executed.
.TP 8
.B -i 
Interactive mode.  After the schematic is read in and parsed and, if specified,
the guile_procedure is execute then go into interactive mode.  Interactive
mode allows the user to execute guile procedures directly. 
.TP 8
schematic1 [... schematicN]
At least one schematic file must be specified.  If multiple schematics are 
specified then they are sequentially read in and parsed with the assumption 
that they are all part of the same design.  It is important that the 
schematic(s) follow all the options (ie last).

.SH EXAMPLES 
These examples assume that you have a stack_1.sch in the current directory.

gnetlist requires that at least one schematic to be specified on the command line:

.nf
	./gnetlist stack_1.sch 
.ad b 

This is not very useful since it does not direct gnetlist to do 
anything.  

Specify a guile procedure name to get gnetlist to output a netlist:

.nf
	./gnetlist -g geda stack_1.sch 
.ad b 

The netlist output will be written to a file called "output.net" 
in the current working directory.

You can specify the output filename by using the -o flag:

.nf
	./gnetlist -g geda stack_1.sch -o stack.netlist
.ad b 

The spice backend is run against the schematic(s) if you specify 
-g spice and the tango backend is run if you specify -g tango.

To interact with the guile interpreter:

.nf
	./gnetlist -i stack_1.sch 
.ad b 

You will get a prompt where you can execute guile procedures.

To get a more verbose feedback as to what gnetlist is doing run 
with the -v flag:

.nf
	./gnetlist -v -g geda stack_1.sch 
.ad b 

.SH "ENVIRONMENT"
No environment variables are used.

.SH "AUTHOR"
Ales V. Hvezda and many others

.SH SEE ALSO
.BR geda (1),
.BR gnetlist (1),
.BR gnetlist (1),
.BR gsymbol (1)
.SH COPYRIGHT
Copyright \(co  1999, 2000, 2001, 2003 Ales V Hvezda

This document can be freely redistributed according to the terms of the 
GNU General Public License version 2.0



NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2014 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 1.9.2
=================================
* Visibility and attributes of several VHDL symbols have been fixed.

* Two obsolete scripts verilog2vhdl.sh and script.sed have been
  removed.

* Undocumented and ambiguous feature of the `spice-sdb' backend for
  gnetlist consisting in the possibility of using the "value="
  attribute instead of "file=" only for include blocks and only in
  embedding mode has been removed.

* The following scripts and programs have been moved from the `utils'
  directory to a new `contrib' directory and are no longer installed
  by default:

         smash_megafile          olib
         gmk_sym                 convert_sym
         sarlacc_schem           sarlacc_sym
         pads_backannotate       sw2asc
         gnet_hier_verilog.sh

* `gpstoimage' has been removed from the distribution; equivalent 
  functionality is provided by the `gaf' command.

* `convert_sym.awk' has been removed from the distribution; equivalent
  functionality is provided by the `convert_sym' command.

* gnetlist now supports output of netlists to the standard output.

Notable changes in gEDA/gaf 1.9.1
=================================

* Unbalanced overbar markers are now rendered correctly.

* Most tools now no longer set the GEDADATA or GEDADATARC environment
  variables (they still obey them though).

* `gschem' now provides a path drawing tool, accessed using <A H> or
  the "Add Path" menu item.

  - Add a cusp node by left-clicking.

  - Add a Bezier curve node by left-dragging from the node point to
    the control point.

  - End the path by clicking on the same node point twice, or clicking
    on the start node to close the path.

* On desktops that reparent menu bars (e.g. Ubuntu Unity) it is once
  again possible to paste the clipboard using the menus.

* A number of `gschem' editing behaviour changes were made:

  - Double-clicking on an entry in the "In Use" tab of the component
    selector now behaves the same way as double-clicking in the
    "Libraries" tab.

  - When new attributes are attached to a net, they are placed at a 50
    unit offset.

  - The text color is now used as the default color for `pinlabel='
    attributes.

  - Pin attributes are no longer automatically replaced when rotating
    or mirroring pins.

  - It is possible to edit a text object even when other non-text
    objects are selected.

  - Control grips are now always drawn at the same on-screen size.

* The multi-attribute editor dialog in `gschem' has been extended to
  support more cases:

  - It now supports editing attributes of multiple selected
    objects. These are aggregated into the dialog's attribute list.

    Multiple attributes with the same name attached to any given
    object are still shown in separate rows, but identically named
    attributes from multiple selected objects will share a row. If a
    row has differing values between objects, its value shows the text
    "<various>" in a muted colour.

  - Where an attribute is present in some, but not all selected
    objects its name in the list is shown in red. A new popup menu
    option allows copying such attributes to all selected objects.

  - If no complex, placeholder, pin, net or bus objects are selected,
    the dialog now falls back to editing any selected attribute text.

    This allows editing of unattached or floating attributes in
    symbols, or any attributes seleccted independantly of their owning
    object. In this fallback mode, all attributes with the same name
    shown in separate rows, and the "add attribute" functionality is
    disabled.

* `gschem' has new icons for many editing commands.

* Several bugs were fixed in `gnetlist':

  - Invalid attributes are ignored.

  - Correct detection of graphical symbols.

  - Components with no `refdes=' attribute are reported.

  - Crash fixes for `drc' and `switcap' backends.

* `gaf export' now accepts a `--scale' option.

* The `--size', `--margins' and `--align' options to `gaf export' can
  now accept their multiple arguments separated with a space ` ', a
  colon `:' or a semicolon `;'.

* `gaf config' now handles the `-p' option properly.

* `gnetlist', `gsymcheck' and `gaf' now output internationalised
  messages.

* New Scheme procedures were added. Please see the "geda-scheme" info
  manual for more details.

  - The (gschem action) module contains procedures for working with
    user editing actions (i.e. commands accessed via the menus or
    keybindings).

  - Two new hooks were added: `bind-keys-hook' is called whenever
    keymap bindings are changed; `action-property-hook' is called when
    action properties are changed.


Notable changes in gEDA/gaf 1.9.0
=================================

* A new schematic renderer library, `libgedacairo' has been added, and
  all tools have been updated to use it.  Related changes:

  - The following configuration functions have been removed:

         output-capstyle         paper-sizes
         output-color            postscript-prolog
         output-orientation      print-command
         output-type             setpagedevice-orientation
         paper-size              setpagedevice-pagesize

  - The following `gschem' configuration items have been added:

         print-paper             print-color
         print-orientation

  - `gschem' now generates PDF files rather than PostScript or EPS
    files.

  - `gschem' now uses a standard GTK print dialog (with printer
    selection, print preview, etc.)

* A new `gaf' command-line utility has been added.  It provides a
  number of subcommands:

  - `gaf export' can be used to convert schematic files to PNG, SVG,

gnetlist/src/gnetlist.c
/* gEDA - GPL Electronic Design Automation
 * gnetlist - gEDA Netlist
 * Copyright (C) 1998-2010 Ales Hvezda
 * Copyright (C) 1998-2010 gEDA Contributors (see ChangeLog for details)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include <config.h>
#include <locale.h>
#include <version.h>

#include <stdio.h>
#include <sys/param.h>
#include <sys/types.h>
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#include <dirent.h>
#ifdef HAVE_STRING_H
#include <string.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif


#include <libgeda/libgeda.h>
#include <libgeda/libgedaguile.h>

#include "../include/globals.h"
#include "../include/prototype.h"
#include "../include/gettext.h"

void gnetlist_quit(void)
{
    s_clib_free();

    /* Free GSList *backend_params */
    g_slist_free (backend_params);
}


void main_prog(void *closure, int argc, char *argv[])
{
    TOPLEVEL *pr_current;

    parse_commandline(argc, argv);

    scm_set_program_arguments (argc, argv, NULL);

    /* this is a kludge to make sure that spice mode gets set */
    /*  Hacked by SDB to allow spice netlisters of arbitrary name
     *        as long as they begin with "spice".  For example, this spice
     *  netlister is valid: "spice-sdb".
     */
    if (guile_proc) {
        if (strncmp(guile_proc, "spice", 5) == 0) {
            netlist_mode = SPICE;
        }
    }

    libgeda_init();

    /* create log file right away */
    /* even if logging is enabled */
    s_log_init ("gnetlist");

    s_log_message(_(
        "gEDA/gnetlist version %s%s.%s\n"
        "gEDA/gnetlist comes with ABSOLUTELY NO WARRANTY; see COPYING for more details.\n"
        "This is free software, and you are welcome to redistribute it under certain\n"
        "conditions; please see the COPYING file for more details.\n\n"),
        PREPEND_VERSION_STRING, PACKAGE_DOTTED_VERSION, PACKAGE_DATE_VERSION);

#if defined(__MINGW32__) && defined(DEBUG)
    fprintf(stderr, _("This is the MINGW32 port.\n\n"));
#endif

    /* register guile (scheme) functions */
    g_register_funcs();
    s_init_traverse ();

    scm_dynwind_begin (0);
    pr_current = s_toplevel_new ();
    edascm_dynwind_toplevel (pr_current);

    /* Evaluate Scheme expressions that need to be run before rc files
     * are loaded. */
    scm_eval (pre_rc_list, scm_current_module ());

    scm_c_use_module ("geda library");

    g_rc_parse (pr_current, argv[0], "gnetlistrc", rc_filename);

    /* immediately setup user params */
    i_vars_init_gnetlist_defaults ();

    /* Load basic gnetlist functions */
    scm_primitive_load_path (scm_from_utf8_string ("gnetlist.scm"));

    scm_dynwind_end();
}

int main(int argc, char *argv[])
{
#if ENABLE_NLS
    setlocale (LC_ALL, "");
    setlocale (LC_NUMERIC, "C");
    bindtextdomain ("geda-gnetlist", LOCALEDIR);
    textdomain ("geda-gnetlist");
    bind_textdomain_codeset("geda-gnetlist", "UTF-8");
#endif
    scm_boot_guile (argc, argv, main_prog, 0);
    return 0;
}
